_posts/2019-07-03-get-to-most-out-of-the-linker-map-file.md

```diff
@@ -0,0 +1,234 @@
+
+```
+    .rodata.main.str1.4
+    	              0x00000000000017b8        0x4 _build/nrf52840_xxaa/main.c.o
+```
+
+I also noticed that the inclusion of `_ctype_` added 0x101 bytes of read-only data in the `text` area 😯.
+
+```
+    .rodata._ctype_
+                    0x00000000000017c0      **0x101** /usr/local/Cellar/arm-none-eabi-gcc/8-2018-q4-major/gcc/bin/../lib/gcc/arm-none-eabi/8.2.1/../../../../arm-none-eabi/lib/thumb/v7e-m+fp/hard/libc_nano.a(lib_a-ctype_.o)
+                    0x00000000000017c0                **_ctype_**
+```
+
+Including a call to `atoi` made the `text` area to be increased but also the `data` area. I used a diff tool to make things simpler and discovered data that was before discarded by the linker:
+
+```
+    .data._impure_ptr
+    								0x0000000000000000        0x4 /usr/local/Cellar/arm-none-eabi-gcc/8-2018-q4-major/gcc/bin/../lib/gcc/arm-none-eabi/8.2.1/../../../../arm-none-eabi/lib/thumb/v7e-m+fp/hard/libc_nano.a(lib_a-impure.o)
+```
+
+One last piece of information to keep in mind: initialized variables have to be kept in Flash but they appear in `RAM` in the map file because they are copied into RAM before entering the `main` function. It means that you better have to initialize variables to `0` to put them in the `BSS` section in order to free up some Flash. Here, symbols `__data_start__` and `__data_end__` keep track of the area used in RAM to keep initialized variables. Those values are stored starting at `0x00000000000018d0`:
+
+```
+    .data           0x0000000020000008       0x70 load address 0x00000000000018d0
+                    0x0000000020000008                __data_start__ = .
+    [...]
+    								0x0000000020000078                __data_end__ = .
+```
+
+Note that symbols that are compiled but not needed are also listed in the map file in the `Discarded input sections` part.
+
+Most of the information from the map file has been analyzed. 
+
+Several usages of the map file are possible. Most of the time, you will have an address and you will want to know which function or data is being used. But some other times, it will be useful for debugging...
+
+## Debugging a linking error
+
+Map files are created while the built code (`.o` files) is linked together which means it can be used to resolve errors in the linking process. I remember working on a bootloader that was contained in a few Flash pages. At some point, I wanted to use `atoi` but the bootloader didn't compile anymore because there were no more space available...
+
+Using the previous example, let's say I now have only 0x800 bytes of Flash. If I want to compile the example that is not using `atoi`, I don't have any issue. But if I add the call to `atoi`, which needs several symbols to be linked into the executable and I obtain:
```

Comment:
~~If I want to compile~~ Compiling the first example, without `atoi`, there are no issues. The second example however would overflow our small flash.